ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -249,20 +252,44 @@ checkAdversarialChain recipe adv = do
                 Nothing     -> pure ()   -- there are no remaining honest active slots
                                          --
                                          -- TODO hpc shows this never executes
-    checkDensity :: Exn.Except (AdversarialViolation hon adv) ()
-    checkDensity =
-        when (C.Count s <= C.windowSize winA) $ do
-            let -- first stability window after the intersection in the honest
-                -- schema
-                hScgWin = C.UnsafeContains (C.Count $ C.getCount $ C.windowStart winA) (C.Count s)
-                -- first stability window after the intersection in the
-                -- alternative schema
-                aScgWin = C.UnsafeContains (C.Count 0) (C.Count s)
-                -- number of active slots in the first stability window after
-                -- the intersection in the honest schema
-                hCount = C.getCount $ BV.countActivesInV S.notInverted (C.sliceV hScgWin vH)
-                aCount = C.getCount $ BV.countActivesInV S.notInverted (C.sliceV aScgWin vA)
-            when (aCount >= hCount) $ Exn.throwError $ BadDensity hCount aCount
+
+    -- | Check that the density of the adversarial schema is less than the
+    -- density of the honest schema in the first stability window after the
+    -- intersection and in any subwindow that contains the first race to the
+    -- k+1st block.
+    --
+    -- See description of @ensureLowerDensityInWindows@
+    checkDensity = do
+        let
+          -- window of the honest schema after the intersection
+          winHAfterIntersection =
```

Comment:
If I understand & recall correctly, I think this could be written `winHAfterIntersection = C.joinWin winH carWin`, if the argument to `checkAdversarialChain` were a _checked_ recipe. (Or perhaps `checkAdversarialChain` could simply start by checking the given recipe?)

Reply:
Currently
```haskell
winHAfterIntersection :: C.Contains SlotE hon ...
```
because the code is using it to slice a `hon` vector.

On the other hand
```haskell
C.joinWin winH carWin :: C.Contains SlotE base adv
```
which doesn't fit.

Reply:
Hmm, I see, thanks.

Actually, is it just `winHAfterIntersection = carWin`?

```
winA :: C.Contains SlotE base adv
winH :: C.Contains SlotE base hon
carWin :: C.Contains SlotE hon adv

winA == C.joinWin winH carWin
```